### DISCUSSION TITLE
About the working of method `MagickConnectedComponentsImage`?

### DISCUSSION THREAD

**Discussion by Haidar0096** [ImageMagick/ImageMagick#5947] [Development] [answered]
### ImageMagick version

7

### Operating system, version and so on

windows 11 64-bit

### Description

I need to understand how does this method work. I do understand that it tries to label connected components in an image.

Its signature is `MagickBooleanType MagickConnectedComponentsImage(MagickWand *wand,
  const size_t connectivity,CCObjectInfo **objects)` so I understand it returns an array of `CCObjectInfo` where each element is a connected components object info struct (`CCObjectInfo`) i.e. it represents an object of connected pixels.

But there is no return value that specifies how many boxes the algorithm found, so how can I know the length of the returned array?

Here is an example of the values I get in the returned array at indices 0, 65535, and 65536:
![image](https://user-images.githubusercontent.com/49842031/211169672-83452873-cf5e-43ef-b32d-3fd848ce5904.png)
![image](https://user-images.githubusercontent.com/49842031/211169697-83b3e3f6-40a6-4860-9468-b699f8f3d2d0.png)
![image](https://user-images.githubusercontent.com/49842031/211169721-f5177c0f-ede6-4285-9274-7ecc6603e80f.png)

**Comment by snibgo** [comment]
> I need to understand how does this method [MagickConnectedComponentsImage] work. 

Many MagickWand functions are just wrappers for MagickCore functions, which do the real work.

MagickConnectedComponentsImage() calls ConnectedComponentsImage(), which is in MagickCore\vision.c. You need to read that function.

ConnectedComponentsImage() returns an image, which MagickConnectedComponentsImage() replaces in the image list. Towards the end of ConnectedComponentsImage(), code writes values from the `object` array, using `i` as an index, numbered from `0` to `component_image->colors`.

**Reply by Haidar0096** [reply]
> MagickConnectedComponentsImage() calls ConnectedComponentsImage(), which is in MagickCore\vision.c. You need to read that function.


I did read both of them, but none of them return the size of the returned array `CCObjectInfo **objects`. So how can I traverse it?

**Comment by snibgo** [comment]
From `colors`, which is a component of the Image structure that ConnectedComponentsImage() returns.

**Reply by Haidar0096** [reply]
> From `colors`, which is a component of the Image structure that ConnectedComponentsImage() returns.

So if I want to use the MagickWand api method, i.e. `MagickConnectedComponentsImage()`, then I can't know the length of the returned array `CCObjectInfo **objects` ? What is the point returning this array from this function in this case?

**Comment by snibgo** [comment, marked as answer]
> So if I want to use the MagickWand api method, i.e. MagickConnectedComponentsImage(), then I can't know the length of the returned array CCObjectInfo **objects ?

Yes you can. The wand contains the modified image. One component of that image is `colors`.

**Reply by Haidar0096** [reply]
okay, so right after calling MagickConnectedComponentsImage, I should call GetImageFromMagickWand 
to get a handle to the same modified image, then I can read its member called colors to get the size of the array.

I expected the size to be passed as a pointer in the first place and then filled with the size of the array.

Thanks for help.

**Reply by Haidar0096** [reply]
@snibgo just to double check, this is the code you meant?
```
	Image* imageBeforeEditing = GetImageFromMagickWand(wand);
	CCObjectInfo* objectInfo = NULL;
	MagickBooleanType res = MagickConnectedComponentsImage(wand, 4, &objectInfo);
	Image* currentImage = GetImageFromMagickWand(wand);
	size_t numColors = currentImage->colors; // this will contain the length of the objectInfo array
```

**Comment by snibgo** [comment]
That looks good, aside from the first line. `imageBeforeEditing` has no meaning after `MagickConnectedComponentsImage`, because the image has been replaced.
